### PR TITLE
feat: chainbase plugin for eliza

### PR DESCRIPTION
**Pull request by lxcong**
# Relates to
Eliza AI agents, blockchain data analytics, Chainbase

# Risks
Low. This is a new feature that adds functionality without disrupting existing systems.

# Background
## What does this PR do? 
This PR introduces a new Chainbase plugin for Eliza that enables AI agents to access and analyze blockchain data across multiple networks using natural language queries. The plugin leverages Chainbase's comprehensive data infrastructure to provide real-time insights.

## What kind of change is this?
Features (non-breaking change which adds functionality)

## Why are we doing this? Any context or related work?
Eliza AI agents currently lack the ability to easily access and analyze on-chain data, especially across multiple blockchain networks. This plugin bridges that gap by allowing users to query blockchain data using natural language without requiring deep technical blockchain knowledge. It unlocks powerful new analytics and insight generation capabilities for Eliza.

This work builds upon Chainbase's existing multi-chain data infrastructure and Eliza's natural language processing capabilities. It is part of the broader initiative to make blockchain data more accessible and actionable for end users.

# Documentation changes needed?
My changes require a change to the project documentation.
I have updated the documentation to include:
- Overview of the Chainbase plugin and its purpose
- Installation and configuration instructions 
- Examples of natural language queries and the blockchain insights generated
- API reference for key functions

# Testing
## Where should a reviewer start?
1. Review the updated documentation to understand the purpose and functionality of the plugin
2. Run the automated test suite: `npm run test`
3. Follow the instructions to install and configure the plugin in a dev environment
4. Test sample natural language queries against blockchain data and verify the insights generated match expected results
5. Review the code changes, focusing on the integration between Eliza's NLP and Chainbase's data layer

## Detailed testing steps
1. Install and configure the Chainbase plugin by following the documentation 
2. From the Eliza interface, input a natural language query such as "What is the total transaction volume on Ethereum in the last 24 hours?" 
3. Verify the plugin converts this to the appropriate Chainbase query and returns the expected analytics
4. Test queries across multiple supported blockchain networks 
5. Verify error handling for unsupported networks or malformed queries
6. Run edge case tests from the automated test suite and ensure code coverage meets standards

# Deployment instructions
1. Merge PR after review and testing approval
2. Update Eliza production environment to point to the new Chainbase plugin version
3. Restart Eliza web servers
4. Smoke test Chainbase queries from the production Eliza interface 
5. Monitor error logs and system metrics post-deployment to ensure stability

Let me know if you need any other information to include in the PR description! I aimed to be thorough while following the provided template.

# Screenshot
image

image

## Summary by CodeRabbit

- **New Features**
	- Added Chainbase plugin for blockchain data querying
	- Introduced Instagram configuration options
	- Enabled token balance retrieval and on-chain data querying

- **Configuration**
	- Added environment variables for Chainbase and Instagram integration
	- Configured API keys and authentication parameters

- **Documentation**
	- Added README for Chainbase Plugin explaining its functionality and usage

### CODE REVIEW

**Comment by daizhengxue** [comment]
great work @lxcong 
Would it be possible to integrate Chainbase data API into other plugins? For example, @plugin-solana is currently using Birdeye for its wallet token list. It would be helpful to have Chainbase data provider as a second option.
Other EVM plugins and blockchain plugins are facing similar issues with having only a single data provider.

**Comment by lxcong** [comment]
> great work @lxcong Would it be possible to integrate Chainbase data API into other plugins? For example, @plugin-solana is currently using Birdeye for its wallet token list. It would be helpful to have Chainbase data provider as a second option. Other EVM plugins and blockchain plugins are facing similar issues with having only a single data provider. ! [image](https://private-user-images.githubusercontent.com/47060721/402313641-03de7983-fb46-4933-a2dd-6ca4f359b098.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2NjczNTQsIm5iZiI6MTczNjY2NzA1NCwicGF0aCI6Ii80NzA2MDcyMS80MDIzMTM2NDEtMDNkZTc5ODMtZmI0Ni00OTMzLWEyZGQtNmNhNGYzNTliMDk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEyVDA3MzA1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhZDk0NDRmOWIxMWI5ZjkwYmQ0NWEyOTkxMGY4ZjRmOGNmNWVhYzE5NzQ4NTUzNmEwYmU4YzRmNDAzZmRiNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K8yCdJ-AQAyVFPXTTIV3i8hPEjtAUSu5LFzN6u1y6FE)

That's a very good suggestion, I will create a new PR to provide a multi-chain data provider.

File: agent/src/index.ts [hunk 1]
```diff
@@ -95,14 +95,15 @@ import { OpacityAdapter } from "@elizaos/plugin-opacity";
...
+import { chainbasePlugin } from "@elizaos/plugin-chainbase";
 import { quaiPlugin } from "@elizaos/plugin-quai";
 import Database from "better-sqlite3";
 import fs from "fs";
 import net from "net";
 import path from "path";
 import { fileURLToPath } from "url";
 import yargs from "yargs";
-import {dominosPlugin} from "@elizaos/plugin-dominos";
+import { dominosPlugin } from "@elizaos/plugin-dominos";
```
Comment (tcm390) [inline agent/src/index.ts:106]: probably should remove this

Comment (lxcong) [reply agent/src/index.ts]: It has been fixed

File: packages/plugin-chainbase/link-to-docs
Comment (HashWarlock) [inline packages/plugin-chainbase/link-to-docs:1]: What's this for?

File: packages/plugin-chainbase/package.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    "version": "0.1.7-alpha.1",
+    "main": "dist/index.js",
+    "type": "module",
+    "types": "dist/index.d.ts",
+    "dependencies": {
+        "@elizaos/core": "workspace:*"
+    },
+    "scripts": {
+        "build": "tsup --format esm --dts",
+        "dev": "tsup --format esm --dts --watch"
```
Comment (HashWarlock) [inline packages/plugin-chainbase/package.json:12]:
```suggestion
        "dev": "tsup --format esm --dts --watch",
        "lint": "eslint --fix  --cache ."
```
Add lint script

File: packages/plugin-chainbase/src/actions/queryData.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            const messageText = message.content.text.toLowerCase();
+
+            if (!messageText.includes(QUERY_PREFIX)) {
+                callback({
+                    text: `Please use the format: ${QUERY_PREFIX} <your natural language query>`,
+                });
+                return;
+            }
+
+            const queryText = message.content.text
```
Comment (HashWarlock) [inline packages/plugin-chainbase/src/actions/queryData.ts:47]: Does this give a better way to extract the query versus using a template to extract the query? I can imagine the text can have anything after the `QUERY_PREFIX` which could include inappropriate text.

I think this is okay, but something to think about in the future.

File: packages/plugin-chainbase/link-to-docs
Comment (lxcong) [reply packages/plugin-chainbase/link-to-docs:1]: Thanks for the help I found the useless file and I cleared it

File: packages/plugin-chainbase/src/actions/queryData.ts (hunk 3 above)
Comment (lxcong) [reply packages/plugin-chainbase/src/actions/queryData.ts:47]: Very good suggestion and I feel like I need to start a new PR to optimize. This includes which prompt words are functionally scoped

Comment (lxcong) [reply packages/plugin-chainbase/src/actions/queryData.ts:47]: Linked to an issue: https://github.com/elizaOS/eliza/issues/2320

**Review by HashWarlock — DISMISSED** [review]
I think there are conflicts to be fixed, but LGTM. Let me know when fixes are resolved then I can help get merged. Sorry on delay

**Comment by lxcong** [comment]
> I think there are conflicts to be fixed, but LGTM. Let me know when fixes are resolved then I can help get merged. Sorry on delay

Sorry, I accidentally merged and closed the PR
I have recreated the PR https://github.com/elizaOS/eliza/pull/2586

File: packages/plugin-chainbase/src/templates/index.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
+export const responsePrompt = (
+    result: any,
+    query: string
+) => `You are a blockchain data analyst. Your task is to analyze and present the query results from Chainbase in a clear and informative way.
+
+SQL Query:"""
+${result.sql}
+"""
```
[the remark this answers is not in the document]
Comment (lxcong) [reply packages/plugin-chainbase/src/templates/index.ts:8]: For presentation

File: packages/plugin-chainbase/src/actions/retrieveTokenBalance.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                contract_address,
+            });
+
+            // Convert hex balance to decimal and adjust for token decimals
+            const processedTokens = tokens.map((token) => ({
+                ...token,
+                balance: token.balance
+                    ? Number(
+                          BigInt(token.balance) / BigInt(10 ** token.decimals)
+                      ).toFixed(18)
```
[the remark this answers is not in the document]
Comment (lxcong) [reply packages/plugin-chainbase/src/actions/retrieveTokenBalance.ts]: good advice

File: packages/plugin-chainbase/src/libs/chainbase.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        const response = await fetch(
+            `${CHAINBASE_API_URL_ENDPOINT}/api/v1/text2sql`,
+            {
+                method: "POST",
+                headers: {
+                    "Content-Type": "application/json",
+                },
+                body: JSON.stringify({
+                    assistant_id: "6b86c502-d203-4f6b-baf6-f406c23a9421",
+                    input: {
```
[the remark this answers is not in the document]
Comment (lxcong) [reply packages/plugin-chainbase/src/libs/chainbase.ts:29]: For the time being, only one Assistant ID is supported

File: packages/plugin-chainbase/src/actions/queryData.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                modelClass: ModelClass.SMALL,
+            });
+
+            callback({
+                text: formattedResponse,
+            });
+        } catch (error) {
+            elizaLogger.error("Error in queryChainbase action:", error);
+            return "error";
+        }
```
[the remark this answers is not in the document]
Comment (lxcong) [reply packages/plugin-chainbase/src/actions/queryData.ts:88]: good advice

File: packages/plugin-chainbase/src/actions/retrieveTokenBalance.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (lxcong) [reply packages/plugin-chainbase/src/actions/retrieveTokenBalance.ts]: updated

File: packages/plugin-chainbase/src/actions/queryData.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (lxcong) [reply packages/plugin-chainbase/src/actions/queryData.ts]: updated

**Comment by lxcong** [comment]
> I think there are conflicts to be fixed, but LGTM. Let me know when fixes are resolved then I can help get merged. Sorry on delay

Conflict has been resolved, please help merge

**Review by HashWarlock — APPROVED** [review]
LGTM
